File: src/clu/sources/ipmitool.py
import logging
import re

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source, PARSE_FAIL_MSG
from clu.input import text_program

log = logging.getLogger(__name__)


class Ipmitool(Source):
    _lan_print_keys = [
        "bmc.ipv4_source",
        "bmc.ipv4_address",
        "bmc.ipv4_mask",
        "bmc.mac_address",
    ]
    _mc_info_keys = [
        "bmc.firmware_version",
        "bmc.manufacturer_id",
        "bmc.manufacturer_name",
    ]

    def provides(self, provides: Provides) -> None:
        for key in self._lan_print_keys + self._mc_info_keys:
            provides[key] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.append("ipmitool")

    def parse(self, facts: Facts) -> None:
        if facts["phy.platform"] != "physical":
            log.info("Not a physical platform, skipping bmc parsing")
            return

        self._parse_ipmitool_mc_info(facts)
        self._parse_ipmitool_lan_print(facts)

    def _parse_ipmitool_lan_print(self, facts: Facts) -> None:
        regexes = {
            r"^IP Address Source *: (.+)": "bmc.ipv4_source",
            r"^IP Address *: (.+)": "bmc.ipv4_address",
            r"^Subnet Mask *: (.+)": "bmc.ipv4_mask",
            r"^MAC Address *: (.+)": "bmc.mac_address",
        }
        fields = {}
        data, rc = text_program("ipmitool lan print")
        if data == "" or rc != 0:
            log.warning("Failed to run ipmitool lan print")
            for name in self._lan_print_keys:
                facts[name] = PARSE_FAIL_MSG
            return

        for regex, field in regexes.items():
            match = re.search(regex, data, re.MULTILINE)
            value = match.group(1).strip() if match else None
            log.debug(f"{value=}")
            if value is not None:
                fields[field] = value
        log.debug(f"{fields=}")

        facts.update(fields)

    def _parse_ipmitool_mc_info(self, facts: Facts) -> None:
        regexes = {
            r"^Firmware Revision *: (.+)": "bmc.firmware_version",
            r"^Manufacturer ID *: (.+)": "bmc.manufacturer_id",
            r"^Manufacturer Name *: (.+)": "bmc.manufacturer_name",
        }
        fields = {}
        data, rc = text_program("ipmitool mc info")
        if data == "" or rc != 0:
            log.warning("Failed to run ipmitool mc info")
            for name in self._mc_info_keys:
                facts[name] = PARSE_FAIL_MSG
            return

        for regex, field in regexes.items():
            match = re.search(regex, data, re.MULTILINE)
            value = match.group(1).strip() if match else None
            log.debug(f"{value=}")
            if value is not None:
                fields[field] = value
        log.debug(f"{fields=}")

        facts.update(fields)

File: internal/facts/sources/ipmitool.go
// SPDX-FileCopyrightText: 2024 Hunter Matthews
// SPDX-License-Identifier: LGPL-2.1-only

package sources

// Go port of src/clu/sources/ipmitool.py
// Collects BMC/IPMI facts via `ipmitool mc info` and `ipmitool lan print`.
// Skips parsing unless platform is physical (facts["phy.platform"] == "physical").

import (
	"regexp"
	"strings"

	"github.com/huntermatthews/clu/internal/facts/types"
	"github.com/huntermatthews/clu/internal/input"
)

// Ipmitool gathers BMC firmware/network details.
type Ipmitool struct{}

var ipmitoolLanFacts = map[string]*types.Fact{
	"bmc.ipv4_source":  {Name: "bmc.ipv4_source", Origin: `(?m)^IP Address Source *: (.+)`, Tier: types.TierTwo},
	"bmc.ipv4_address": {Name: "bmc.ipv4_address", Origin: `(?m)^IP Address *: (.+)`, Tier: types.TierOne},
	"bmc.ipv4_mask":    {Name: "bmc.ipv4_mask", Origin: `(?m)^Subnet Mask *: (.+)`, Tier: types.TierTwo},
	"bmc.mac_address":  {Name: "bmc.mac_address", Origin: `(?m)^MAC Address *: (.+)`, Tier: types.TierOne},
}
var ipmitoolMcInfoFacts = map[string]*types.Fact{
	"bmc.firmware_version":  {Name: "bmc.firmware_version", Origin: `(?m)^Firmware Revision *: (.+)`, Tier: types.TierOne},
	"bmc.manufacturer_id":   {Name: "bmc.manufacturer_id", Origin: `(?m)^Manufacturer ID *: (.+)`, Tier: types.TierTwo},
	"bmc.manufacturer_name": {Name: "bmc.manufacturer_name", Origin: `(?m)^Manufacturer Name *: (.+)`, Tier: types.TierTwo},
}

// Provides registers all fact keys produced by this source.
func (i *Ipmitool) Provides(p types.Provides) {
	for name := range ipmitoolLanFacts {
		p[name] = i
	}
	for name := range ipmitoolMcInfoFacts {
		p[name] = i
	}
}

// Requires declares external program dependency.
func (i *Ipmitool) Requires(r *types.Requires) {
	r.Programs = append(r.Programs, "ipmitool")
}

// Parse orchestrates the two ipmitool queries if platform is physical.
func (i *Ipmitool) Parse(f *types.FactDB) {
	platform, _ := f.Get("phy.platform")
	if platform != "physical" { // skip if not physical hardware
		return
	}
	i.parseMcInfo(f)
	i.parseLanPrint(f)
}

func (i *Ipmitool) parseLanPrint(f *types.FactDB) {
	for _, fact := range ipmitoolLanFacts {
		fact.Value = types.ParseFailMsg
	}
	defer f.AddAll(ipmitoolLanFacts)

	data, rc, _ := input.CommandRunner("ipmitool lan print")
	if data == "" || rc != 0 {
		return
	}
	for _, fact := range ipmitoolLanFacts {
		re := regexp.MustCompile(fact.Origin)
		if m := re.FindStringSubmatch(data); len(m) == 2 {
			if val := strings.TrimSpace(m[1]); val != "" {
				fact.Value = val
			}
		}
	}
}

func (i *Ipmitool) parseMcInfo(f *types.FactDB) {
	for _, fact := range ipmitoolMcInfoFacts {
		fact.Value = types.ParseFailMsg
	}
	defer f.AddAll(ipmitoolMcInfoFacts)

	data, rc, _ := input.CommandRunner("ipmitool mc info")
	if data == "" || rc != 0 {
		return
	}
	for _, fact := range ipmitoolMcInfoFacts {
		re := regexp.MustCompile(fact.Origin)
		if m := re.FindStringSubmatch(data); len(m) == 2 {
			if val := strings.TrimSpace(m[1]); val != "" {
				fact.Value = val
			}
		}
	}
}
